src/main/java/com/emc/pravega/perf/WriterWorker.java

```diff
@@ -0,0 +1,111 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.emc.pravega.perf;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Random;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ *  abstract class for Writers.
+ */
+public abstract class WriterWorker extends Worker implements Callable<Void> {
+
+    WriterWorker(int sensorId, int eventsPerSec, int secondsToRun,
+                 boolean isRandomKey, int messageSize, Instant start,
+                 PerfStats stats, String streamName, long totalEvents) {
+
+        super(sensorId, eventsPerSec, secondsToRun,
+                isRandomKey, messageSize, start,
+                stats, streamName, totalEvents, 0);
+    }
+
+    /**
+     * writes the data.
+     *
+     * @param key            key for data.
+     * @param data           data to write
+     */
+    public abstract CompletableFuture writeData(String key, String data);
+
+    /**
+     * flush the producer data.
+     *
+     */
+    public abstract void flush();
+
+    /**
+     * Increment the Event counter and return the updated value
+     *  @return incremented value.
+     *
+     */
+
+    public abstract long eventCountIncrementAndGet();
+
+    /**
+     *  get the current event counter value
+     *  @return current value.
+     *
+     */
+    public abstract long eventCountGet();
+
+    @Override
+    public Void call() throws InterruptedException, ExecutionException, IllegalStateException {
+        CompletableFuture retFuture = null;
+
+        do {
+            final Instant loopStartTime = Instant.now();
+            for (int i = 0; (i < eventsPerSec) && (eventCountIncrementAndGet() <= totalEvents) &&
```

Comment:
Why do the children need to implement this?
eventCountIncrementAndGet() is an instance method that is defined above but it is being compared against a totalEvents count.
If this is meant to be an instance variable, then it can simply be made into a local variable because it is not used in the child class.
If it is meant to be a global, then why is the child forced to implement it? there is only one possible implementation.

Reply:
removed